pkg/driver/trace_driver.go

```diff
@@ -634,34 +491,73 @@ func (d *Driver) internalRun(iatOnly bool, generated bool) {
 		allRecordsWritten.Wait()
 	}
 
+	statSuccess := atomic.LoadInt64(&successfulInvocations)
+	statFailed := atomic.LoadInt64(&failedInvocations)
+
 	log.Infof("Trace has finished executing function invocation driver\n")
-	log.Infof("Number of successful invocations: \t%d\n", atomic.LoadInt64(&successfulInvocations))
-	log.Infof("Number of failed invocations: \t%d\n", atomic.LoadInt64(&failedInvocations))
+	log.Infof("Number of successful invocations: \t%d", statSuccess)
+	log.Infof("Number of failed invocations: \t%d", statFailed)
+	log.Infof("Total invocations: \t\t\t%d", statSuccess+statFailed)
+	log.Infof("Failure rate: \t\t\t%.2f%%", float64(statFailed)*100.0/float64(statSuccess+statFailed))
 }
 
-func (d *Driver) RunExperiment(iatOnly bool, generated bool) {
-	if iatOnly {
-		log.Info("Generating IAT and runtime specifications for all the functions")
-		for i, function := range d.Configuration.Functions {
-			spec := d.SpecificationGenerator.GenerateInvocationData(
-				function,
-				d.Configuration.IATDistribution,
-				d.Configuration.ShiftIAT,
-				d.Configuration.TraceGranularity,
-			)
-			d.Configuration.Functions[i].Specification = spec
+func (d *Driver) GenerateSpecification() {
+	log.Info("Generating IAT and runtime specifications for all the functions")
 
-			file, _ := json.MarshalIndent(spec, "", " ")
-			err := os.WriteFile("iat"+strconv.Itoa(i)+".json", file, 0644)
-			if err != nil {
-				log.Fatalf("Writing the loader config file failed: %s", err)
-			}
+	for i, function := range d.Configuration.Functions {
+		// Equalising all the InvocationStats to the first function
+		if d.Configuration.LoaderConfiguration.DAGMode {
+			function.InvocationStats.Invocations = d.Configuration.Functions[0].InvocationStats.Invocations
 		}
+		spec := d.SpecificationGenerator.GenerateInvocationData(
+			function,
+			d.Configuration.IATDistribution,
+			d.Configuration.ShiftIAT,
+			d.Configuration.TraceGranularity,
+		)
+
+		d.Configuration.Functions[i].Specification = spec
+	}
+}
+
+func (d *Driver) outputIATsToFile() {
+	for i, function := range d.Configuration.Functions {
+		file, _ := json.MarshalIndent(function.Specification, "", " ")
+		err := os.WriteFile("iat"+strconv.Itoa(i)+".json", file, 0644)
+		if err != nil {
+			log.Fatalf("Writing the loader config file failed: %s", err)
+		}
+	}
+}
+
+func (d *Driver) DumpSpecification(writeIATsToFile bool, readIATsFromFile bool) {
```

Comment:
This function does unexpected thing: it acts as an input as well. Base on the name, I would expect it to be only output.

Reply:
Fixed.